android-core/src/main/java/com/mparticle/internal/DeviceAttributes.kt

```diff
@@ -0,0 +1,343 @@
+                            MPUtility.AdIdInfo.Advertiser.GOOGLE -> deviceInfo.put(MessageKey.GOOGLE_ADV_ID, adIdInfo.id)
+                        }
+                        message = "Successfully collected " + adIdInfo.advertiser.descriptiveName + " Advertising ID."
+                    }
+                }
+            } catch (jse: JSONException) {
+                Logger.debug("Failed while building device-customAttributes object: ", jse.toString())
+            }
+        }
+        if (firstCollection) {
+            Logger.debug(message)
+            firstCollection = false
+        }
+
+        try {
+            val mParticle = MParticle.getInstance()
+            if (mParticle != null) {
+                val configManager = mParticle.Internal().configManager
+                val registration = configManager.pushRegistration
+                if (registration != null && !MPUtility.isEmpty(registration.instanceId)) {
+                    deviceInfo.put(MessageKey.PUSH_TOKEN, registration.instanceId)
+                    deviceInfo.put(MessageKey.PUSH_TOKEN_TYPE, Constants.GOOGLE_GCM)
+                }
+            }
+        } catch (jse: JSONException) {
+            Logger.debug("Failed while building device-customAttributes object: ", jse.toString())
+        }
+    }
+
+    fun getDeviceInfo(context: Context): JSONObject {
+        if (deviceInfo == null) {
+            deviceInfo = getStaticDeviceInfo(context)
+        }
+        deviceInfo?.let {
+            updateDeviceInfo(context, it)
+            return it
+        } ?: run {
+            return JSONObject()
+        }
+    }
+
+    fun getAppInfo(context: Context): JSONObject {
+        return getAppInfo(context, false)
+    }
+
+    //https://go.mparticle.com/work/SQDSDKS-6951
+    fun getAppInfo(context: Context, forceUpdateInstallReferrer: Boolean): JSONObject {
+        if (appInfo == null) {
+            appInfo = getStaticApplicationInfo(context)
+            appInfo?.let { updateInstallReferrer(context, it) }
+        } else if (forceUpdateInstallReferrer) {
+            appInfo?.let { updateInstallReferrer(context, it) }
+        }
+        return appInfo as JSONObject
+    }
+
+    val operatingSystemString: String
+        get() = when (operatingSystem) {
+            MParticle.OperatingSystem.ANDROID -> Constants.Platform.ANDROID
+            MParticle.OperatingSystem.FIRE_OS -> Constants.Platform.FIRE_OS
+            else -> Constants.Platform.ANDROID
+        }
+
+    companion object {
+        // re-use this whenever an attribute can't be determined
+        const val UNKNOWN: String = "unknown"
+
+        @Volatile
+        private var _deviceImei: String? = null
+
+        @get:JvmStatic
+        val deviceImei: String?
+            get() = _deviceImei ?: null
+
+        @JvmStatic
+        fun setDeviceImei(deviceImei: String?) {
+            _deviceImei = deviceImei
+        }
```

Comment:
Can this just be rewritten as a single property? What's the idea behind having a get only property with a private backing ivar and then a separate setter method?

Reply:
Since it’s a single property and needed to be changed in the existing method,So I created separate getter and setter for it.